### PR TITLE
fix(input, input-number): increment/decrement unsafe numbers without loss of precision

### PR DESCRIPTION
**Pull request by benelan**
**Related Issue:** #5920

## Summary

Numbers that cannot be represented in JavaScript's IEEE-754 double precision were being rounded when incrementing/decrementing because they were cast as Number (we store the value as a string). We use BigDecimal class to support numbers that do not fit into JavaScript's encoding method. I had overlooked the increment/decrement functionality when implementing BigDecimal.

BigDecimal uses BigInt internally, and unfortunately BigInt doesn't support exponential notation. For example, `BigInt("1e2") ` throws the error:

> Uncaught SyntaxError: invalid BigInt syntax

To work around the constraint, I manually expand any exponential notation numbers into normal decimal numbers. All of this is done via string manipulation, so sorry in advance for the burned retinas you are about to receive from this PR review.

For more info:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger#description

### CODE REVIEW

File: src/utils/number.ts [hunk 1]
```diff
@@ -52,6 +55,10 @@
     return `${this.isNegative ? "-" : ""}${integers}${decimals.length ? "." + decimals : ""}`;
   }
 
+  toFixed(decimalPlaces: number): string {
+    return this.toString().replace(new RegExp(`-?\d+\.?\d{0, ${decimalPlaces}}`), "");
```
[the remark this answers is not in the document]
Comment (benelan) [reply src/utils/number.ts:59]: Huh, TIL. 

After looking into this, I don't think we even need to use toFixed() anymore. It was originally added in #3473 to deal with Javascript's precision issues, but using BigDecimal is a much more robust approach. I'll see if nuking the scary regex breaks everything or not

File: src/components/input-number/input-number.tsx [hunk 2]
```diff
@@ -751,7 +761,7 @@ export class InputNumber
     event.stopPropagation();
   };
 
-  private setChildNumberElRef = (el) => {
+  private setChildNumberElRef = (el: HTMLInputElement) => {
```
Comment (jcfranco) [inline src/components/input-number/input-number.tsx:764]: ✨🏆✨

File: src/utils/math.ts [hunk 3]
```diff
@@ -1,7 +1,9 @@
 export const clamp = (value: number, min: number, max: number): number => Math.max(min, Math.min(value, max));
 
-export const decimalPlaces = (value: number): number => {
-  const match = ("" + value).match(/(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/);
+const decimalNumberRegex = new RegExp(/(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/);
+
+export const decimalPlaces = (value: string): number => {
+  const match = value.match(decimalNumberRegex);
```
Comment (jcfranco) [inline src/utils/math.ts:6]: Suggestion: would `test` help simplify this?

File: src/components/tooltip/tooltip.e2e.ts [hunk 4]
```diff
@@ -12,6 +12,8 @@ describe("calcite-tooltip", () => {
   /**
    * Helps assert the canceled Esc key press when closing tooltips
    * Must be called before the tooltip is closed via keyboard.
+   *
+   * @param page
```
Comment (jcfranco) [inline src/components/tooltip/tooltip.e2e.ts:16]: Can you roll this file back? I think it was auto-fixed by the linter, but the file is unrelated to this PR.

File: src/components/slider/slider.tsx [hunk 5]
```diff
@@ -1200,7 +1200,7 @@ export class Slider
     const percent = (x - left) / width;
     const mirror = this.shouldMirror();
     const clampedValue = this.clamp(this.min + range * (mirror ? 1 - percent : percent));
-    let value = Number(clampedValue.toFixed(decimalPlaces(this.step)));
+    let value = Number(clampedValue.toFixed(decimalPlaces(`${this.step}`)));
```
Comment (jcfranco) [inline src/components/slider/slider.tsx:1203]: If this fixes precision on the slider, can you update the PR title to reflect this? If it's to fix types, can you submit this in a separate PR? If it's the latter, we can merge the separate PR first and we can skip rolling it back here.

File: src/utils/number.ts [hunk 6]
```diff
@@ -157,6 +166,52 @@ export function sanitizeExponentialNumberString(numberString: string, func: (s:
...
+ *
+ * @param {string} numberString - pre-validated exponential or decimal number
+ * @returns {string} numberString in decimal notation
+ */
+export function expandExponentialNumberString(numberString: string): string {
+  const exponentialParts = numberString.split(/[eE]/);
+  if (exponentialParts.length === 1) {
+    return numberString;
+  }
+  if (Number.isSafeInteger(+numberString)) {
```
Comment (jcfranco) [inline src/utils/number.ts:181]: Nitpick: for readability, can you add a newline between if statements?

Comment (jcfranco) [inline src/utils/number.ts:181]: Can you store `+numberString` in a local variable and reuse it?

File: src/components/input/input.tsx [hunk 7]
```diff
@@ -610,26 +615,32 @@ export class Input
...
+      nudgedValue.subtract(`${inputMin}`).isNegative;
+
+    const nudgedValueAboveInputMax =
+      typeof inputMax === "number" &&
+      !isNaN(inputMax) &&
+      !nudgedValue.subtract(`${inputMax}`).isNegative;
 
-    const inputValPlaces = decimalPlaces(inputVal);
-    const inputStepPlaces = decimalPlaces(inputStep);
+    const finalValue = nudgedValueBelowInputMin
```
Comment (jcfranco) [inline src/components/input/input.tsx:633]: Suggestion: can this be refactored to compute and evaluate a nudged value check at a time? It's currently computing two values where one might possibly be thrown away.

File: src/components/slider/slider.tsx (hunk 5 above)
Comment (benelan) [reply src/components/slider/slider.tsx:1203]: The parameter needs to be a string so the decimalPlaces function can also be used with BigDecimal. Previously `step` was being converted to a string in the decimalPlaces function itself, but we would lose precision if converting BigDecimal to `number` before passing it to the function. So it doesn't fix anything in slider, it's just a small, required refactor.

File: src/utils/math.ts (hunk 3 above)
Comment (benelan) [reply src/utils/math.ts:6]: We need to use the values of the match a couple lines down

File: src/components/slider/slider.tsx (hunk 5 above)
Comment (benelan) [reply src/components/slider/slider.tsx:1203]: Nevermind, Input isn't using decimalPlaces anymore, removing!
